### PR TITLE
Use read/write trait to simplify codec

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @siddontang PTAL

### CODE REVIEW

**Comment by ngaut** [comment]
PTAL @coocood @shenli @qiuyesuifeng

File: src/util/codec/number.rs [hunk 1]
```diff
@@ -185,120 +181,113 @@ mod test {
...
+                o => panic!("expect {:?}, got {:?}", $k, o),
+            }
         }
-        panic!("out_of_bound should be thrown!");
     }
 
-    fn test_decode_out_of_bound<B, R, F: Fn(B) -> Result<R>>(f: F, buf: B) {
-        if let Err(Error::OutOfBound(_, _)) = f(buf) {
-            return;
+    macro_rules! test_eof {
```
Comment (ngaut) [inline src/util/codec/number.rs]: Please add comments for macro

**Comment by BusyJay** [comment]
ping

**Comment by coocood** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Comment by ngaut** [comment]
LGTM

**Pull request merged**